Ceph: Assure mon canary pods deletion

jmolmo opened the pull request:
**Description of your changes:**
~With this modification the mon failover process is not executed if there is
no available nodes where to schedule the mon pod.
This avoid to have mon canary pods in Pending state because no nodes
available.~

If for any cause the operator processing a mon failover is restarted, the
mon canary pods created in order to get the right placement for this
kind of pods can remain in <pending> state forever.

The following labels have been added to the mon canary deployments:
"mon_type" = "canary"
"canary_pvc_name" = <PVC name used for the canary pod>

A separate <removeCanaryDeployments> has been implemented for
removing the dependence of the variable <canaryCleanup>.
This function uses the <canary> labels to obtain the <real>
and <existing> list of mon canary deployments in any moment.

This change, by itself, will avoid most of the problems with
pending canary pods, but as second safe guard measure a call to this
function has been added in the health check, to <force> a canary pod
deletion if the cluster of monitors is healthy and complete (this implies
that no canary pods should be present)

Resolves #3764

[test ceph]

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

Resolves #3764
 
**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.


**Details about manual testing of this fix:**
Lab used: (4 vms centos 7)

```
# kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:12:17Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}

[root@rook-ceph-operator-77b4b868d8-7jgx6 /]# rook version
rook: v1.1.0-beta.0.661.gdca96729-dirty
go: go1.12.7
```

In first place check wher the operator is running:

```
# kubectl -n rook-ceph describe pod rook-ceph-operator-844c546bb5-rdwps  | grep Node:
Node:         ku-worker-00.karmalabs.com/192.168.122.132
```

Take a look to the vm list:

```
# kcli list vms
Using local libvirt as no client was specified...
+--------------+--------+-----------------+------------------------------------+-----------------+---------+--------+
|     Name     | Status |       Ips       |               Source               |       Plan      | Profile | Report |
+--------------+--------+-----------------+------------------------------------+-----------------+---------+--------+
| ku-master-00 |   up   |  192.168.122.40 | CentOS-7-x86_64-GenericCloud.qcow2 | agitated_muster |  kvirt  |        |
| ku-worker-00 |   up   | 192.168.122.132 | CentOS-7-x86_64-GenericCloud.qcow2 | agitated_muster |  kvirt  |        |
| ku-worker-01 |   up   |  192.168.122.67 | CentOS-7-x86_64-GenericCloud.qcow2 | agitated_muster |  kvirt  |        |
| ku-worker-02 |   up   | 192.168.122.131 | CentOS-7-x86_64-GenericCloud.qcow2 | agitated_muster |  kvirt  |        |
+--------------+--------+-----------------+------------------------------------+-----------------+---------+--------+

```

Stopping a node different from the one where operator is running:

```
# date && kcli stop vm ku-worker-02
jue dic 19 09:41:03 CET 2019
Using local libvirt as no client was specified...
Stopping vm ku-worker-02 in local...
ku-worker-02 stopped
```

In the Operator logs we can see:

```
2019-12-19 09:41:21.676813 W | op-mon: mon c not found in quorum, waiting for timeout before failover
2019-12-19 09:41:52.164022 W | op-mon: mon c not found in quorum, waiting for timeout before failover
2019-12-19 09:42:22.644601 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
```

Once the timeout for the mon is expired, the failover is tried, but is not possible because we do not have enough Nodes "ready"

```
2019-12-19 09:42:22.647447 E | op-mon: Not enough <Ready> nodes(only 3) to schedule the mon "c". It is needed at least 4 nodes. Canceling failover
2019-12-19 09:42:53.113920 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-19 09:42:53.117656 E | op-mon: Not enough <Ready> nodes(only 3) to schedule the mon "c". It is needed at least 4 nodes. Canceling failover
2019-12-19 09:43:23.572786 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-19 09:43:23.575372 E | op-mon: Not enough <Ready> nodes(only 3) to schedule the mon "c". It is needed at least 4 nodes. Canceling failover
2019-12-19 09:43:54.045977 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-19 09:43:54.048877 E | op-mon: Not enough <Ready> nodes(only 3) to schedule the mon "c". It is needed at least 4 nodes. Canceling failover
2019-12-19 09:44:24.572873 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-19 09:44:24.575831 E | op-mon: Not enough <Ready> nodes(only 3) to schedule the mon "c". It is needed at least 4 nodes. Canceling failover
2019-12-19 09:44:55.045390 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
....
```

This will continue while the "lost node" situation persist. If we check the pod status we will see several pods in "Terminating" state, these pods belongs to the stopped node.
Besides that we have a mon "pending" pod. This pending pods has been created by k8s (trying to satisfy the mon deployment and it is scheduled in the stopped node).

```
# kubectl -n rook-ceph get pods
NAME                                                              READY   STATUS        RESTARTS   AGE
csi-cephfsplugin-6cdrp                                            3/3     Running       3          72m
...
csi-cephfsplugin-provisioner-56c8b7ddf4-bj8f8                     4/4     Terminating   0          72m
...
rook-ceph-crashcollector-ku-master-00.karmalabs.com-f4b4d6zlskh   1/1     Running       0          24h
...
rook-ceph-crashcollector-ku-worker-02.karmalabs.com-5895bdkgrsj   1/1     Terminating   0          24h
rook-ceph-crashcollector-ku-worker-02.karmalabs.com-5895bdws8sq   0/1     Pending       0          94s
rook-ceph-mgr-a-57996b5dcc-kstkn                                  1/1     Running       0          24m
rook-ceph-mon-a-5db6888d7b-nvxhv                                  1/1     Running       0          44m
rook-ceph-mon-b-5b84574448-njdg4                                  1/1     Running       0          19m
rook-ceph-mon-c-7d99bcbf5-bf9dz                                   1/1     Terminating   0          24h
rook-ceph-mon-c-7d99bcbf5-w9gp2                                   0/1     Pending       0          94s
rook-ceph-mon-d-8467ff498d-sfxqq                                  1/1     Running       0          24h
rook-ceph-operator-844c546bb5-rdwps                               1/1     Running       0          10m
rook-ceph-osd-0-547ff9fd4d-4dlfl                                  1/1     Running       0          19m
rook-ceph-osd-1-fd4dbcbcf-6jnzt                                   1/1     Running       3          24h
rook-ceph-osd-2-57b95ff546-5v22c                                  1/1     Terminating   1          24h
rook-ceph-osd-2-57b95ff546-ntrss                                  0/1     Pending       0          94s
rook-ceph-osd-3-749ccd686f-ngn9k                                  1/1     Running       0          44m
...

```

The "canary" mon pod is not scheduled/launched because we do not have enough Nodes.

If we start again the stopped k8s node:

```
# date && kcli start vm ku-worker-02
jue dic 19 09:52:08 CET 2019
Using local libvirt as no client was specified...
Starting vm ku-worker-02...
ku-worker-02 started
```

in the operator logs we can see:

```
2019-12-19 09:53:04.111704 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-19 09:53:04.116220 I | op-mon: Failing over monitor "c"
2019-12-19 09:53:04.116271 I | op-mon: starting new mon: &{ResourceName:rook-ceph-mon-e DaemonName:e PublicIP: Port:6789 DataPathMap:0xc000a034a0}
2019-12-19 09:53:04.122530 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-e-canary
2019-12-19 09:53:04.127777 I | op-mon: sched-mon: waiting for canary pod creation rook-ceph-mon-e-canary
```

Here we can see two different scenarios:

1. If the node stopped never returns to live: it will be needed to add a new k8s Node. The normal operator mon failover procedure will be executed and the "lost" monitor will be scheduled in the completely new node.

2. If the node stopped returns to live:
Kubernetes will restart by itself the "lost" monitor in the "restarted" node. In fact, this mon pod was scheduled (due the mon deployment) in the moment that the node was stopped and remains in "Pending" state during the time the node is stopped. What we can see in the pods state is something like:

```
# kubectl -n rook-ceph get pods
...
rook-ceph-mon-b-78d8b7cff8-jlp8w                                  1/1     Terminating   0          144m
rook-ceph-mon-b-78d8b7cff8-qn5hk                                  0/1     Pending       0          37s
...

```
And the "Pending" mon pod is in this state because the deployment of this pod requires the node that is stopped:

```
# kubectl -n rook-ceph describe pod rook-ceph-mon-b-78d8b7cff8-qn5hk

Events:
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  0/4 nodes are available: 1 node(s) had taints that the pod didn't tolerate, 3 node(s) didn't match node selector.
  Warning  FailedScheduling  <unknown>  default-scheduler  0/4 nodes are available: 1 node(s) had taints that the pod didn't tolerate, 3 node(s) didn't match node selector.

```

Once the node is recovered k8s take two actions, 1 delete the old mon pod in "Terminating" state and the other one is to continue with the start of the mon pod in "Pending" state. This allows to recover the mon without the intervention of the operator(mon health check). In this case what we can see in the operator log is:

```
2019-12-19 13:56:15.896037 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-19 13:56:15.898963 E | op-mon: Not enough <Ready> nodes(only 3) to schedule the mon "c". It is needed at least 4 nodes. Canceling failover
2019-12-19 13:57:01.354828 I | op-mon: mon c is back in quorum, removed from mon out timeout list
```

[test ceph]

mykaul commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -196,8 +197,17 @@ func (c *Cluster) failMon(monCount, desiredMonCount int, name string) {
-			logger.Errorf("failed to failover mon %q. %v", name, err)
+		// only possible if we have enough "ready nodes"
+		if readyNodes, err := k8sutil.GetKubernetesReadyNodes(c.context.Clientset); err != nil {
+			logger.Errorf("list of <Ready> nodes for the mon %q not available: %v", name, err)
```
nit: remove 'the'

BlaineEXE replied:
Rook uses periods to separate errors from messages.

```suggestion
			logger.Errorf("list of <Ready> nodes for the mon %q not available. %v", name, err)
```

BlaineEXE replied:
I worry that `not available` might imply there is not an explicit failure here when that is actually the case. I would suggest `failed to get list of <Ready> nodes for mon %q` instead. This follows the `failed to...` pattern present in a lot of Rook's code.

mykaul commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -196,8 +197,17 @@ func (c *Cluster) failMon(monCount, desiredMonCount int, name string) {
+			logger.Errorf("list of <Ready> nodes for the mon %q not available: %v", name, err)
+		} else {
+			if len(readyNodes) < desiredMonCount {
+				logger.Errorf("Not enough <Ready> nodes(only %d) to schedule the mon %q. It is needed at least %d nodes. Canceling failover", len(readyNodes), name, desiredMonCount)
```
nit: "It is needed at least %d nodes" -> "At least %d nodes are needed"

BlaineEXE replied:
Error messages should not start with capitals. Also, more detailed reasoning should be later in the description; I think the ordering could be flipped.
```suggestion
				logger.Errorf("caneling failover. not enough <Ready> nodes (%d ready, %d needed) to schedule mon %q", len(readyNodes), desiredMonCount, name)
```

BlaineEXE commented on pkg/operator/k8sutil/node.go:
```diff
@@ -364,3 +364,21 @@ func GenerateNodeAffinity(nodeAffinity string) (*v1.NodeAffinity, error) {
+	var readyNodes []v1.Node
+	for _, node := range nodes.Items {
+		for _, nodeCondition := range node.Status.Conditions {
+			if nodeCondition.Type == v1.NodeReady && nodeCondition.Status == "True" {
```
@travisn is this PR merely recreating the placement logic that Noah worked so hard to get rid of? The point of the canary pods is to let k8s tell Rook whether mons can be scheduled instead of implementing a bunch of custom logic in Rook that duplicates what k8s does. Some basic sanity checks seem okay to me, but I would advise against the tempation to try to "optimize" too much here.

BlaineEXE replied:
That sounds much better to me. I didn't realize we weren't cleaning up canaries if they were unschedulable.

jmolmo replied:
@BlaineEXE, @travisn : 

Well, really what we have had all the time is more or less the same that @travisn  has described as possible solution.
@noahdesu  did an excellent work here and this situation is also covered. The canary-pod is deleted regularly, and a new canary pod is created as consequence of the restart of the failover procedure. This cycle is infinite, and for that reason it seems that a canary pod is in pending state forever( it is created, deleted in each failover cycle).
I think that this behavior is correct, and in any case the only thing what we can do is to reduce the time checking if the canary pod is ok or not (reducing canaryRetries var). 

The solution i have implemented in this PR is to avoid the failover procedure when we know in advance that there are no nodes available to complete this process successfully. And therefore, the pending mon canary pod (different instances...) will be always present.

What should i do? 
- Should i continue with this solution improving the implementation taking into account also the mon placement? 

- It does not matter to have the pending canary pod being recreated all the time? should we close #3764  as  "won't fix"?
Note: If there are Nodes available, the mon canary pod is always deleted, or at least i cannot be able to reproduce the situation where the mon canary pod is not deleted.

**Details about the failover cycle with an example:**

Once the timeout period for acquire quorum pass the failover for the mon starts:
```

2019-12-20 08:46:51.450088 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
2019-12-20 08:46:51.450113 I | op-mon: Failing over monitor "c"
2019-12-20 08:46:51.450138 I | op-mon: starting new mon: &{ResourceName:rook-ceph-mon-e DaemonName:e PublicIP: Port:6789 DataPathMap:0xc000ff0750}
2019-12-20 08:46:51.454986 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-e-canary
2019-12-20 08:46:51.474624 I | op-mon: sched-mon: waiting for canary pod creation rook-ceph-mon-e-canary
```

Here we are in the situation where the canary pod is created, but it cannot be scheduled because there are no nodes available. 

We are 15 minutes waiting because we are retrying the mon-canary pod creation for 
canaryRetries(180) * canaryRetryDelaySeconds(5) = 900 seconds = 15 minutes

Once we have exhausted the number of retries, the canary pod is cleaned:

```
2019-12-20 09:01:53.512780 I | op-mon: assignmon: cleaning up canary deployment rook-ceph-mon-e-canary and canary pvc 
2019-12-20 09:01:53.512832 I | op-k8sutil: removing deployment rook-ceph-mon-e-canary if it exists
2019-12-20 09:01:53.520543 I | op-k8sutil: Removed deployment rook-ceph-mon-e-canary
2019-12-20 09:01:53.526424 I | op-k8sutil: rook-ceph-mon-e-canary still found. waiting...
2019-12-20 09:01:55.531978 I | op-k8sutil: confirmed rook-ceph-mon-e-canary does not exist

```
And a message reports the situation:

`2019-12-20 09:01:55.532043 E | op-mon: failed to failover mon "c". failed to place new mon on a node: assignmon: error scheduling monitor: sched-mon: canary pod scheduling failed retries
`

And a new failover cycle starts when the mon health check is launched again (each <HealthCheckInterval> = 45 seconds)

`
2019-12-20 09:02:41.001187 W | op-mon: mon c NOT found in quorum and timeout exceeded, mon will be failed over
`

jmolmo replied:
@travisn: probably wong link: ( [this one](https://rook-io.slack.com/archives/CK9CF5H2R/p1576862700036500) ) ... encrypted OSDs?

BlaineEXE commented:
Removing `reliability` label from this, as I don't think this PR makes Rook more reliable. Instead, I view it as an optimization.

jmolmo commented:
Finally i was able to reproduce the issue!!.:-)  It happens if the operator processing the mon failover is restarted.
@BlaineEXE, @travisn  : Changed completely the implementation based in your suggestions.

jmolmo commented:
Brief summary of the current behaviour:

Lab: 4 nodes (master + 3 workers) . Ceph cluster with 4 mons.

Test:

- Stop one of the worker nodes ( different from the one with the operator)

- wait until the mon failover happens, we will see in the log:

```
2020-01-03 08:22:10.588500 W | op-mon: mon b not found in quorum, waiting for timeout before failover
...
2020-01-03 08:22:56.072571 W | op-mon: mon b not found in quorum, waiting for timeout before failover
...
2020-01-03 08:23:41.598076 W | op-mon: mon b not found in quorum, waiting for timeout before failover
...
```

- k8s tries to start again the mon in the stopped node, but no resources, so the pod remains in "pending" state.

```
rook-ceph-mon-b-5fd9f94544-p45mp                                  1/1     Terminating   0          16m
rook-ceph-mon-b-5fd9f94544-pzbp8                                  0/1     Pending       0          53s
```


- failover happens:

```
2020-01-03 08:31:16.654459 I | op-mon: Failing over monitor "b"
2020-01-03 08:31:16.654472 I | op-mon: starting new mon: &{ResourceName:rook-ceph-mon-e DaemonName:e PublicIP: Port:6789 DataPathMap:0xc0007a8d20}
2020-01-03 08:31:16.654505 D | op-mon: monConfig: &{ResourceName:rook-ceph-mon-e DaemonName:e PublicIP: Port:6789 DataPathMap:0xc0007a8d20}
2020-01-03 08:31:16.665407 I | op-mon: sched-mon: created canary deployment rook-ceph-mon-e-canary
2020-01-03 08:31:16.680492 I | op-mon: sched-mon: waiting for canary pod creation rook-ceph-mon-e-canary
```

- But the cluster does not have enough nodes to run the canary pod:

```
2020-01-03 08:31:21.684902 D | op-mon: sched-mon: monitor rook-ceph-mon-e-canary canary pod rook-ceph-mon-e-canary-76cd76c65f-s9w7p not yet scheduled
2020-01-03 08:31:26.693447 D | op-mon: sched-mon: monitor rook-ceph-mon-e-canary canary pod rook-ceph-mon-e-canary-76cd76c65f-s9w7p not yet scheduled
2020-01-03 08:31:31.699182 D | op-mon: sched-mon: monitor rook-ceph-mon-e-canary canary pod rook-ceph-mon-e-canary-76cd76c65f-s9w7p not yet scheduled

```

- So the Pod is in "Pending" state:

`rook-ceph-mon-e-canary-76cd76c65f-s9w7p                           0/1     Pending       0          48s`


- Restart the operator (scale down to 0 and scale up to 1): The pending canary pod is removed: (this is what didn't happen before this fix)

`2020-01-03 08:33:16.199288 I | op-mon: removeCanaryDeployments: cleaning up canary deployment rook-ceph-mon-e-canary and canary pvc `

- The operator does not create another canary pod for the stopped mon. So no new failover is started ( that is good!), We can see in the log file:

`2020-01-03 08:37:45.771921 D | op-k8sutil: deployment rook-ceph-mon-b status={ObservedGeneration:4 Replicas:1 UpdatedReplicas:1 ReadyReplicas:0 AvailableReplicas:0 UnavailableReplicas:1 Conditions:[{Type:Progressing Status:True LastUpdateTime:2020-01-03 08:11:01 +0000 UTC LastTransitionTime:2020-01-03 08:10:54 +0000 UTC Reason:NewReplicaSetAvailable Message:ReplicaSet "rook-ceph-mon-b-5fd9f94544" has successfully progressed.} {Type:Available Status:False LastUpdateTime:2020-01-03 08:20:59 +0000 UTC LastTransitionTime:2020-01-03 08:20:59 +0000 UTC Reason:MinimumReplicasUnavailable Message:Deployment does not have minimum availability.}] CollisionCount:<nil>}`

- If the node stopped is started again, the pods in "terminating" state are cleared and the pending monitor is started. returning all to the "normal "state"

leseb commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -184,6 +184,12 @@ func (c *Cluster) checkHealth() error {
 		}
 	}
 
+	//Remove any pending/not needed mon canary deployment if everything is ok
```
```suggestion
	// Remove any pending/not needed mon canary deployment if everything is ok
```

jmolmo replied:
thx and sorry for forget always this things .... lets see if i can put any kind of linter/tool to avoid this errors.

leseb commented on pkg/operator/ceph/cluster/mon/health.go:
```diff
@@ -184,6 +184,12 @@ func (c *Cluster) checkHealth() error {
 
+	//Remove any pending/not needed mon canary deployment if everything is ok
+	if allMonsInQuorum && len(quorumStatus.MonMap.Mons) == desiredMonCount {
+		logger.Debugf("mon cluster is healthy, removing any existing canary deployment")
```
```suggestion
		logger.Debug("mon cluster is healthy, removing any existing canary deployment")
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
-				if err := k8sutil.DeleteDeployment(c.context.Clientset, c.Namespace, result.CanaryDeployment); err != nil {
-					logger.Infof("assignmon: error deleting canary deployment %s. %v", result.CanaryDeployment, err)
-				}
+// Delete mon canary deployments (and associated pvcs) using deployment labels
```
```suggestion
// Delete mon canary deployments (and associated PVCs) using deployment labels
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
-					logger.Infof("assignmon: error deleting canary deployment %s. %v", result.CanaryDeployment, err)
-				}
+// Delete mon canary deployments (and associated pvcs) using deployment labels
+// to select this kind of temporal deployments
```
```suggestion
// to select this kind of temporary deployments
```

?

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+// Delete mon canary deployments (and associated pvcs) using deployment labels
+// to select this kind of temporal deployments
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon, mon_type=canary")
```
I feel like this `"app=rook-ceph-mon, mon_type=canary"` should be a variable or a constant...

jmolmo replied:
Only used here... if it is needed in the future in other functions we can use a constant.

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+		logger.Errorf("error trying to get the list of canary deployments: %v", err)
+	} else {
+		for _, canary := range canaryDeployments.Items {
+			logger.Infof("removeCanaryDeployments: cleaning up canary deployment %s and canary pvc %s", canary.Name, canary.Labels["canary_pvc_name"])
```
```suggestion
			logger.Infof("removeCanaryDeployments: cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["canary_pvc_name"])
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+		for _, canary := range canaryDeployments.Items {
+			logger.Infof("removeCanaryDeployments: cleaning up canary deployment %s and canary pvc %s", canary.Name, canary.Labels["canary_pvc_name"])
+			if err := k8sutil.DeleteDeployment(c.context.Clientset, c.Namespace, canary.Name); err != nil {
+				logger.Infof("removeCanaryDeployments: error deleting canary deployment %s. %v", canary.Name, err)
```
```suggestion
				logger.Infof("removeCanaryDeployments: error deleting canary deployment %q. %v", canary.Name, err)
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+				err := c.context.Clientset.CoreV1().PersistentVolumeClaims(c.Namespace).Delete(canary.Labels["canary_pvc_name"], options)
 				if err != nil {
-					logger.Infof("assignmon: error removing canary monitor %s pvc %s. %v", result.CanaryDeployment, result.CanaryPVC, err)
+					logger.Infof("removeCanaryDeployments: error removing canary monitor %s pvc %s. %v", canary.Name, canary.Labels["canary_pvc_name"], err)
```
```suggestion
					logger.Infof("removeCanaryDeployments: error removing canary monitor %q pvc %q. %v", canary.Name, canary.Labels["canary_pvc_name"], err)
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+
+func (c *Cluster) assignMons(mons []*monConfig) error {
+	// when monitors are scheduling below by invoking scheduleMonitor() a canary
+	// deployment and optional canary PVC are created. in order for the
```
```suggestion
	// deployment and optional canary PVC are created. In order for the
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	// when monitors are scheduling below by invoking scheduleMonitor() a canary
+	// deployment and optional canary PVC are created. in order for the
+	// anti-affinity rules to be effective, we leave the canary pods in place
+	// until all of the canaries have been scheduled. only after the
```
```suggestion
	// until all of the canaries have been scheduled. Only after the
```

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -413,6 +413,7 @@ func realScheduleMonitor(c *Cluster, mon *monConfig) (SchedulingResult, error) {
 	d := c.makeDeployment(mon)
 	d.Name += "-canary"
 	d.Spec.Template.ObjectMeta.Name += "-canary"
+	d.Labels["mon_type"] = "canary"
```
I think this label should be applied to both the deployment and the deployment's pod template.

jmolmo replied:
I do not understands well @BlaineEXE. Do you mean, to include this labels directly in the cluster "makeDeployment" method?

BlaineEXE replied:
This is only putting labels on the deployment. The deployment's pod spec template should (I think) also get the label.

jmolmo replied:
I think that is better to leave the code in the spec.go file only for management of "normal" monitors. 
In the realScheduleMonitor function we are labeling the canary monitors only. Normal monitors do not need the "mon_type" label or the "canary_pvc_name". 
What you suggest could be useful if in the case that we have more that "normal" and "canary" mons.

BlaineEXE replied:
I think there is a misunderstanding somewhere. Kubernetes deployments have labels, and the pods they create have a separate spec for labels. If we are labeling the canary deployments, we should also label the deployment's pods via the pod spec.

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -445,6 +446,7 @@ func realScheduleMonitor(c *Cluster, mon *monConfig) (SchedulingResult, error) {
 		_, err = c.context.Clientset.CoreV1().PersistentVolumeClaims(c.Namespace).Create(pvc)
 		if err == nil {
 			pvcName = pvc.Name
+			d.Labels["canary_pvc_name"] = pvcName
```
Should this label also be applied to both the deployment and the deployment's pod template.

jmolmo replied:
Same as said for label "mon_type"

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+// Delete mon canary deployments (and associated PVCs) using deployment labels
+// to select this kind of temporary deployments
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon, mon_type=canary")
```
I have never seen an example of a label selector in the k8s docs with a space after the comma. I'm not sure this will work.
```suggestion
	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon,mon_type=canary")
```

jmolmo replied:
Removed the extra blank space. And, yes...  it works , at least the mon canary deployments are selected properly.:-)

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon, mon_type=canary")
+	if err != nil {
+		logger.Errorf("error trying to get the list of canary deployments: %v", err)
```
Do not use a colon `:` to separate error from message text.

I would also suggest adding here some more context about what Rook is trying to achieve when getting the list of canaries. 

```suggestion
		logger.Errorf("cannot remove canary deployments. error trying to get the list of canary deployments. %v", err)
```

jmolmo replied:
Included your suggestion. thx!.

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon, mon_type=canary")
+	if err != nil {
+		logger.Errorf("error trying to get the list of canary deployments: %v", err)
+	} else {
```
Why not make the above `if err != nil` clause fail fast with a `return` after the `logger.Errorf` so that this `else` block is not necessary?

jmolmo replied:
Agree. More clean. thx!

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+		logger.Errorf("error trying to get the list of canary deployments: %v", err)
+	} else {
+		for _, canary := range canaryDeployments.Items {
+			logger.Infof("removeCanaryDeployments: cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["canary_pvc_name"])
```
Rook doesn't elsewhere use the convention of prefacing logs with the function name.
```suggestion
			logger.Infof("cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["canary_pvc_name"])
```

jmolmo replied:
True. Removed the function name from the log message.

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+		for _, canary := range canaryDeployments.Items {
+			logger.Infof("removeCanaryDeployments: cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["canary_pvc_name"])
+			if err := k8sutil.DeleteDeployment(c.context.Clientset, c.Namespace, canary.Name); err != nil {
+				logger.Infof("removeCanaryDeployments: error deleting canary deployment %q. %v", canary.Name, err)
```
Why `Infof` for an error?

jmolmo replied:
Not using Errorf is my error ... thx!

BlaineEXE commented on pkg/operator/k8sutil/node.go:
```diff
@@ -364,3 +364,21 @@ func GenerateNodeAffinity(nodeAffinity string) (*v1.NodeAffinity, error) {
 	return newNodeAffinity, nil
 }
+
+func GetKubernetesReadyNodes(clientset kubernetes.Interface) ([]v1.Node, error) {
```
Is this addition vestigial now?

jmolmo replied:
yes. I implemented this function in the previous solution approach..., but now is not used.  
Taking into account that it is placed in a general utility package,  maybe is ok to leave it for future use. Do you prefer to delete it?

BlaineEXE replied:
Deleting it seems best since we don't want to encourage others to start reimplementing k8s scheduling logic, and this seems like the foundation of that path.

jmolmo replied:
Deleted

BlaineEXE commented on pkg/operator/k8sutil/node_test.go:
```diff
@@ -389,3 +389,46 @@ func TestGenerateNodeAffinity(t *testing.T) {
 	}
 }
+
+func TestGetKubernetesReadyNodes(t *testing.T) {
```
Vestigial?

jmolmo replied:
removed

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon,mon_type=canary")
+	if err != nil {
+		logger.Errorf("error trying to get the list of canary deployments: %v", err)
+	} else {
```
done. thx!

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,40 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon,mon_type=canary")
+	if err != nil {
+		logger.Errorf("error trying to get the list of canary deployments: %v", err)
```
changed.

jmolmo commented:
Following @travisn suggestion i have changed the label captions to: "mon_canary" and "pvc_name".
Following @BlaineEXE and @travisn suggestions  i have added the labels also to the deployment and to the pod spec template of the deployment.

jmolmo commented on pkg/operator/ceph/cluster/mon/spec.go:
```diff
@@ -51,6 +51,8 @@ func (c *Cluster) getLabels(daemonName string) map[string]string {
 	labels := opspec.PodLabels(AppName, c.Namespace, "mon", daemonName)
 	// Add "mon_cluster: <namespace>" for legacy
 	labels[monClusterAttr] = c.Namespace
+	labels["mon_canary"] = ""
```
getLabel is only called in this file (not exportable). We only know if we are going to create a canary pod in the "mon.go" file. And i think that is simple to set with the rigth value this labels in the exact moment we need them.

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -445,6 +446,7 @@ func realScheduleMonitor(c *Cluster, mon *monConfig) (SchedulingResult, error) {
 		_, err = c.context.Clientset.CoreV1().PersistentVolumeClaims(c.Namespace).Create(pvc)
 		if err == nil {
 			pvcName = pvc.Name
+			d.Labels["pvc_name"] = pvcName
```
Now  I am using the getLabels method, so inserted the labels in all the places needed. I prefered not to add parameters to this function. Because is used only in the spec.go file , and because is easy to set the labels in mon.go when is required

leseb replied:
It makes sense not to change `getLabels()` if this label is only applied for the mons.

jmolmo replied:
and what do you prefer to do @leseb?

leseb replied:
If @BlaineEXE and @travisn are asking for it then please go ahead and address their request.

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,42 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+			}
+		}
+	}
+
```
done

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+			options := &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod, PropagationPolicy: &propagation}
+			err := c.context.Clientset.CoreV1().PersistentVolumeClaims(c.Namespace).Delete(canary.Labels["pvc_name"], options)
+			if err != nil {
+				logger.Infof("removeCanaryDeployments: error removing canary monitor %q pvc %q. %v", canary.Name, canary.Labels["pvc_name"], err)
```
```suggestion
				logger.Infof("failed to delete canary monitor %q pvc %q. %v", canary.Name, canary.Labels["pvc_name"], err)
```

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon,mon_canary=true")
+	if err != nil {
+		logger.Warningf("error trying to get the list of canary deployments: %v", err)
```
changed

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon,mon_canary=true")
+	if err != nil {
+		logger.Warningf("error trying to get the list of canary deployments: %v", err)
```
changed

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+func (c *Cluster) removeCanaryDeployments() {
+	canaryDeployments, err := k8sutil.GetDeployments(c.context.Clientset, c.Namespace, "app=rook-ceph-mon,mon_canary=true")
+	if err != nil {
+		logger.Warningf("failed to get the list of canary deployments: %v", err)
```
```suggestion
		logger.Warningf("failed to get the list of monitor canary deployments. %v", err)
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	for _, canary := range canaryDeployments.Items {
+		logger.Infof("removeCanaryDeployments: cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["pvc_name"])
+		if err := k8sutil.DeleteDeployment(c.context.Clientset, c.Namespace, canary.Name); err != nil {
+			logger.Infof("failed to delete canary deployment %q. %v", canary.Name, err)
```
```suggestion
			logger.Infof("failed to delete canary monitor deployment %q. %v", canary.Name, err)
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	}
+
+	for _, canary := range canaryDeployments.Items {
+		logger.Infof("removeCanaryDeployments: cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["pvc_name"])
```
```suggestion
		logger.Infof("cleaning up canary monitor deployment %q and canary pvc %q.", canary.Name, canary.Labels["pvc_name"])
```

jmolmo commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	for _, canary := range canaryDeployments.Items {
+		logger.Infof("removeCanaryDeployments: cleaning up canary deployment %q and canary pvc %q", canary.Name, canary.Labels["pvc_name"])
+		if err := k8sutil.DeleteDeployment(c.context.Clientset, c.Namespace, canary.Name); err != nil {
+			logger.Infof("removeCanaryDeployments: error deleting canary deployment %q. %v", canary.Name, err)
```
changed

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+	for _, canary := range canaryDeployments.Items {
+		logger.Infof("cleaning up canary monitor deployment %q and canary pvc %q.", canary.Name, canary.Labels["pvc_name"])
+		if err := k8sutil.DeleteDeployment(c.context.Clientset, c.Namespace, canary.Name); err != nil {
+			logger.Infof("failed to delete canary monitor deployment %q. %v", canary.Name, err)
```
```suggestion
			logger.Warningf("failed to delete canary monitor deployment %q. %v", canary.Name, err)
```

leseb commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -551,33 +553,41 @@ func (c *Cluster) initMonIPs(mons []*monConfig) error {
+			options := &metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod, PropagationPolicy: &propagation}
+			err := c.context.Clientset.CoreV1().PersistentVolumeClaims(c.Namespace).Delete(canary.Labels["pvc_name"], options)
+			if err != nil {
+				logger.Infof("failed to delete canary monitor %q pvc %q. %v", canary.Name, canary.Labels["pvc_name"], err)
```
```suggestion
				logger.Warningf("failed to delete canary monitor %q pvc %q. %v", canary.Name, canary.Labels["pvc_name"], err)
```

jmolmo replied:
Yep!!! indeed more coherent ... Thank you !!!

leseb commented:
@BlaineEXE PTAL for a final review. Thanks.

BlaineEXE commented on pkg/operator/ceph/cluster/mon/mon.go:
```diff
@@ -413,6 +413,7 @@ func realScheduleMonitor(c *Cluster, mon *monConfig) (SchedulingResult, error) {
 	d := c.makeDeployment(mon)
 	d.Name += "-canary"
 	d.Spec.Template.ObjectMeta.Name += "-canary"
+	d.Labels["mon_canary"] = "true"
```
Labels are still not applied to the pod spec template per both my and Travis' requests.

jmolmo replied:
@BlaineEXE : The labes are applied, now they are in the `getlabels` function in the spec file.

BlaineEXE replied:
I don't see any changes in this PR or in existing master code where `mon_canary=true` or `pvc_name=XXXXX` is set in `getLabels`. These labels need applied to both deployment and pod spec.

jmolmo replied:
Maybe i don't understand well your request. What i understood was that it was better to have this labels in in the deployment and in the pod spec, and i put this labels in the "getlabels" function to achieve this. What is not set is a value for this labels because the "mon_canary"="true" or "pvc_name" = "whatever" should only be set in the case of the canary pods. 
Do you want to set a default value in this labels?

jmolmo replied:
@BlaineEXE , take into account that the name of the PVC only is known when it is created....is possible to pass this information to the getLabels function, but i think that the code is more clear, readable and easy of understand if we assign directly the values to the labels in the moment that we need it. 
This is the same that it was done previously with:

https://github.com/rook/rook/blob/112a746c625d951678bad80a39e6df3517936cbd/pkg/operator/ceph/cluster/mon/mon.go#L414-L415

Apart of this, If we set this labels for the mon canary pods in the getLabels function, we are changing a function that can be used for any mon, in a function that only can be used for canary pods. Yes, we can add parameters, but as i said previously ... i think that this makes the code more cryptic. (at less for me)

jmolmo commented:
Changed to include the setting of the new "canary" labels in the getLabels function.
Extra: Added  information  about the time left to mon failover in the health check

leseb commented:
@BlaineEXE @travisn PTAL. It'd be nice to merge this today. Thanks.

BlaineEXE commented on pkg/operator/ceph/cluster/mon/spec.go:
```diff
@@ -44,33 +45,35 @@ const (
 	labels := opspec.PodLabels(AppName, c.Namespace, "mon", daemonName)
 	// Add "mon_cluster: <namespace>" for legacy
 	labels[monClusterAttr] = c.Namespace
+	labels["mon_canary"] = strconv.FormatBool(canary)
```
This isn't a big thing, but I think I did prefer the previous usage where the label `mon_canary=true` when the deployment was a canary and where the label was totally absent when it was not a canary. I think it makes it easier to see when something is a canary at a glance as a human. 

```suggestion
	if canary {
		labels["mon_canary"] = true
	}
```

jmolmo replied:
and what abut `pvc_name`? Should it be present only if it has a defined value?

jmolmo replied:
done! :-)
